src/main/kotlin/graphics/scenery/DFTParser.kt

```diff
@@ -0,0 +1,132 @@
+                }
+                else->  {
+                    if (counter == 6)
+                    {
+                        unitCellDimensions[0] = (gridSpacings[0]*gridDimensions[0])+unitCellOrigin[0]
+                        unitCellDimensions[1] = (gridSpacings[1]*gridDimensions[1])+unitCellOrigin[1]
+                        unitCellDimensions[2] = (gridSpacings[2]*gridDimensions[2])+unitCellOrigin[2]
+                        electronicDensity = Array(gridDimensions[0], { Array(gridDimensions[1],
+                            { FloatArray(gridDimensions[2]) } ) } )
+                    }
+                    // Parsing atomic positions.
+                    if (counter < 6+numberOfAtoms){
+                        atomicPositions[counter-6] = Vector3f((line.trim().split("\\s+".toRegex())[2]).toFloat(),
+                                                              (line.trim().split("\\s+".toRegex())[3]).toFloat(),
+                                                              (line.trim().split("\\s+".toRegex())[4]).toFloat())
+                    }
+
+                    // Parsing volumetric data.
+                    // A possible optimization here would be to read this into a 1D array. We cannot directly
+                    // read it into the byte buffer, because we don't know max/min values a-priori.
+                    if (counter >= 6+numberOfAtoms) {
+                        for (value in (line.trim().split("\\s+".toRegex()))) {
+                            // Cube files should be in Fortran (z-fastest ordering).
+                            electronicDensity[xcounter][ycounter][zcounter] = (value).toFloat()
+                            if (electronicDensity[xcounter][ycounter][zcounter] > maxDensity) {
+                                maxDensity = electronicDensity[xcounter][ycounter][zcounter]
+                            }
+                            if (electronicDensity[xcounter][ycounter][zcounter] < minDensity) {
+                                minDensity = electronicDensity[xcounter][ycounter][zcounter]
+                            }
+
+                            zcounter++
+                            if (zcounter == gridDimensions[2]) {
+                                zcounter = 0
+                                ycounter++
+                                if (ycounter == gridDimensions[1]) {
+                                    ycounter = 0
+                                    xcounter++
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+            counter++
+        }
+        // Converting to byte buffer.
+        counter = 0
+        electronicDensityUInt =  MemoryUtil.memAlloc((gridDimensions[0]*
+            gridDimensions[1]*gridDimensions[2]*1))
+        if (normalizeDensityTo > 0.0f){
+            maxDensity = normalizeDensityTo
+        }
+        for (z in 0 until gridDimensions[2]){
```

Comment:
Flat indexing might work better here 👍

Reply:
Done!

Reply:
(and I agree, that's a good idea. Using a 3D array was more to eliminate additional sources of error when coding up the parser, now that I understand what's going where, the switch makes sense).